# 🧪 Z-Test in Statistics
### Hypothesis Testing using Z-Score |

---

> **Z-Test** is used to determine whether the **sample mean** is significantly different from the **population mean**, when the population standard deviation is **known**.

---

## 📌 Problem Statement

A researcher claims that the **average height of adults** in a city is **170 cm**.
We collect a sample of **10 individuals** and want to test if this claim holds true.

| Detail | Value |
|---|---|
| Population Mean (μ) | 170 cm |
| Population Std Dev (σ) | 3 cm |
| Sample Size (n) | 10 |
| Significance Level (α) | 0.05 |

## 🔢 Z-Test Formula

$$Z = \frac{\bar{x} - \mu}{\sigma / \sqrt{n}}$$

- $\bar{x}$ = Sample Mean  
- $\mu$ = Population Mean  
- $\sigma$ = Population Standard Deviation  
- $n$ = Sample Size

In [2]:
# Importing required libraries
import numpy as np
from scipy.stats import norm

## 📋 Step 1 — Define Sample & Population Parameters

We define our collected sample of heights along with the known population parameters.

In [3]:
sample = [172, 174, 168, 169, 171, 173, 175, 170, 169, 172]
population_mean = 170
population_std = 3

samp_mean = np.mean(sample)
n = len(sample)

print(f"Sample       : {sample}")
print(f"Sample Mean  : {samp_mean}")
print(f"Pop Mean     : {population_mean}")
print(f"Pop Std Dev  : {population_std}")
print(f"Sample Size  : {n}")

Sample       : [172, 174, 168, 169, 171, 173, 175, 170, 169, 172]
Sample Mean  : 171.3
Pop Mean     : 170
Pop Std Dev  : 3
Sample Size  : 10


## 🔍 Step 2 — Calculate the Z-Score

The Z-Score tells us how many standard deviations our sample mean is away from the population mean.

$$Z = \frac{\bar{x} - \mu}{\sigma / \sqrt{n}}$$

In [4]:
z_score = (samp_mean - population_mean) / (population_std / np.sqrt(n))

print(f"Z-Score : {z_score:.4f}")

Z-Score : 1.3703


## 📊 Step 3 — Calculate the P-Value

The **P-Value** is the probability of getting a result as extreme as ours, assuming H₀ is true.

We use a **two-tailed test** because we're checking if the mean is *different* (not just higher or lower).

> Formula: `p_value = 2 * (1 - norm.cdf(|Z|))`

In [5]:
p_value = 2 * (1 - norm.cdf(abs(z_score)))

print(f"P-Value : {p_value:.6f}")

P-Value : 0.170587


## ✅ Step 4 — Decision: Accept or Reject Null Hypothesis?

### Our Hypotheses:
- **H₀ (Null Hypothesis):** Average height = 170 cm *(no significant difference)*
- **H₁ (Alternative Hypothesis):** Average height ≠ 170 cm *(significant difference exists)*

### Decision Rule:

| Condition | Decision |
|---|---|
| p_value < α (0.05) | ❌ Reject H₀ |
| p_value ≥ α (0.05) | ✅ Accept H₀ |

In [6]:
alpha = 0.05

print(f"Z-Score  : {z_score:.4f}")
print(f"P-Value  : {p_value:.6f}")
print(f"Alpha    : {alpha}")
print("-" * 40)

if p_value < alpha:
    print("❌ Reject the Null Hypothesis")
    print("→ Sample mean is significantly different from population mean.")
else:
    print("✅ Accept the Null Hypothesis")
    print("→ No significant difference found. The claim holds true.")

Z-Score  : 1.3703
P-Value  : 0.170587
Alpha    : 0.05
----------------------------------------
✅ Accept the Null Hypothesis
→ No significant difference found. The claim holds true.


## 📝 Conclusion

| Metric | Value |
|---|---|
| Sample Mean | 171.3 cm |
| Population Mean | 170 cm |
| Z-Score | 1.3703 |
| P-Value | 0.1706 |
| Alpha (α) | 0.05 |
| Decision | ✅ Accept H₀ |

---

**Interpretation:**
Since **p-value (0.1706) > alpha (0.05)**, we **fail to reject** the null hypothesis.
There is **not enough statistical evidence** to say the average height differs from 170 cm.

> 💡 **Key Takeaway:** A high p-value does NOT prove H₀ is true — it simply means we don't have enough evidence to reject it.

